crates/bellpepper-sha512/src/sha512.rs

```diff
@@ -0,0 +1,506 @@
+        s1 = s1.xor(cs.namespace(|| "first xor for s1"), &new_e.rotr(18))?;
+        s1 = s1.xor(cs.namespace(|| "second xor for s1"), &new_e.rotr(41))?;
+
+        // ch := (e and f) xor ((not e) and g)
+        let ch = UInt64::sha512_ch(cs.namespace(|| "ch"), &new_e, &f, &g)?;
+
+        // temp1 := h + S1 + ch + k[i] + w[i]
+        let temp1 = vec![
+            h.clone(),
+            s1,
+            ch,
+            UInt64::constant(ROUND_CONSTANTS[i]),
+            w[i].clone(),
+        ];
+
+        // S0 := (a rightrotate 28) xor (a rightrotate 34) xor (a rightrotate 39)
+        let new_a = a.compute(cs.namespace(|| "deferred a computation"), &[])?;
+        let mut s0 = new_a.rotr(28);
+        s0 = s0.xor(cs.namespace(|| "first xor for s0"), &new_a.rotr(34))?;
+        s0 = s0.xor(cs.namespace(|| "second xor for s0"), &new_a.rotr(39))?;
+
+        // maj := (a and b) xor (a and c) xor (b and c)
+        let maj = UInt64::sha512_maj(cs.namespace(|| "maj"), &new_a, &b, &c)?;
+
+        // temp2 := S0 + maj
+        let temp2 = vec![s0, maj];
+
+        /*
+        h := g
+        g := f
+        f := e
+        e := d + temp1
+        d := c
+        c := b
+        b := a
+        a := temp1 + temp2
+        */
+
+        h = g;
+        g = f;
+        f = new_e;
+        e = Maybe::Deferred(temp1.iter().cloned().chain(Some(d)).collect::<Vec<_>>());
+        d = c;
+        c = b;
+        b = new_a;
+        a = Maybe::Deferred(
+            temp1
+                .iter()
+                .cloned()
+                .chain(temp2.iter().cloned())
+                .collect::<Vec<_>>(),
+        );
+    }
+
+    /*
+        Add the compressed chunk to the current hash value:
+        h0 := h0 + a
+        h1 := h1 + b
+        h2 := h2 + c
+        h3 := h3 + d
+        h4 := h4 + e
+        h5 := h5 + f
+        h6 := h6 + g
+        h7 := h7 + h
+    */
+
+    let h0 = a.compute(
+        cs.namespace(|| "deferred h0 computation"),
+        &[current_hash_value[0].clone()],
+    )?;
+
+    let h1 = UInt64::addmany(
+        cs.namespace(|| "new h1"),
+        &[current_hash_value[1].clone(), b],
+    )?;
+
+    let h2 = UInt64::addmany(
+        cs.namespace(|| "new h2"),
+        &[current_hash_value[2].clone(), c],
+    )?;
+
+    let h3 = UInt64::addmany(
+        cs.namespace(|| "new h3"),
+        &[current_hash_value[3].clone(), d],
+    )?;
+
+    let h4 = e.compute(
+        cs.namespace(|| "deferred h4 computation"),
+        &[current_hash_value[4].clone()],
+    )?;
+
+    let h5 = UInt64::addmany(
+        cs.namespace(|| "new h5"),
+        &[current_hash_value[5].clone(), f],
+    )?;
+
+    let h6 = UInt64::addmany(
+        cs.namespace(|| "new h6"),
+        &[current_hash_value[6].clone(), g],
+    )?;
+
+    let h7 = UInt64::addmany(
+        cs.namespace(|| "new h7"),
+        &[current_hash_value[7].clone(), h],
+    )?;
+
+    Ok(vec![h0, h1, h2, h3, h4, h5, h6, h7])
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
The tests which run quickly would benefit from running on more than 1 input, consider adding, like in the u64 tests, a
```rust
for _ in 1..100 {
   <test content>
}
```
inner loop. Indicative runtimes in opt mode:
```
        PASS [   0.156s] bellpepper-sha512 sha512::test::test_hash
        PASS [   0.164s] bellpepper-sha512 sha512::test::test_full_block
        PASS [   0.236s] bellpepper-sha512 uint64::test::test_uint64_sha512_maj
        PASS [   0.241s] bellpepper-sha512 uint64::test::test_uint64_sha512_ch
        PASS [   0.297s] bellpepper-sha512 sha512::test::test_full_hash
        PASS [  29.519s] bellpepper-sha512 sha512::test::test_against_vectors
```

See https://github.com/lurk-lab/bellpepper-gadgets/pull/8 for how those numbers are relevant (setting an optimization level of 3 for CI)